### PR TITLE
Metadata

### PR DESCRIPTION
**Pull request by rowanG077**
Moved metadata to json.

TODOs:

- [x] Make old test pass
- [x] Make new tests
- [x] Rewrite packages depending on irmaseal-core to use new API.

### CODE REVIEW

File: irmaseal-core/src/meta.rs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+/// Metadata which contains the version
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub struct Metadata {
+    pub version: Version,
+    pub media_type: ArrayString<[u8; 255]>,
+    pub media_metadata: serde_json::Value,
+    #[serde(serialize_with = "to_base64_ciphertext", deserialize_with = "from_base64_ciphertext")]
+    pub ciphertext: ArrayVec<[u8; CIPHERTEXT_SIZE]>,
+    #[serde(serialize_with = "to_base64_iv", deserialize_with = "from_base64_iv")]
+    pub iv: ArrayVec<[u8; IVSIZE]>,
```
Comment (ivard) [inline irmaseal-core/src/meta.rs:24]: `iv` has a static size of exact 16 bytes and `ciphertext` of exact 144 bytes. This means that an ArrayVec is not necessary for this, it can just be an array with a static length. Can serde handle this?

File: irmaseal-core/src/meta.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+#[serde(rename_all = "SCREAMING_SNAKE_CASE")]
+pub enum Version {
+    V1_0,
+}
+
+/// Metadata which contains the version
+#[derive(Serialize, Deserialize, Debug, PartialEq)]
+pub struct Metadata {
+    pub version: Version,
+    pub media_type: ArrayString<[u8; 255]>,
```
Comment (ivard) [inline irmaseal-core/src/meta.rs:19]: Is it practical to switch to the std version of string now we dropped the requirement to use `nostd`?

File: irmaseal-core/src/meta.rs [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    pub media_metadata: serde_json::Value,
+    #[serde(serialize_with = "to_base64_ciphertext", deserialize_with = "from_base64_ciphertext")]
+    pub ciphertext: ArrayVec<[u8; CIPHERTEXT_SIZE]>,
+    #[serde(serialize_with = "to_base64_iv", deserialize_with = "from_base64_iv")]
+    pub iv: ArrayVec<[u8; IVSIZE]>,
+    pub identity: Identity,
+}
+
+// Can't have natural at the type level(yet). so just use to_base64 twice with different sizes.
+fn to_base64_ciphertext<S>(ciphertext: &ArrayVec<[u8; CIPHERTEXT_SIZE]>, serializer: S) -> Result<S::Ok, S::Error>
```
Comment (ivard) [inline irmaseal-core/src/meta.rs:29]: If you use a fixed size array for `iv` and `ciphertext`, and these fixed size arrays all have a `as_slice` method. Maybe serde understands this such that you can make one general function for both?

File: irmaseal-core/src/meta.rs (hunk 1 above)
Comment (rowanG077) [reply irmaseal-core/src/meta.rs:31]: What's the difference between an ArrayVec and a normal array? No where in the code base normal arrays are used so I avoided it as well.

File: irmaseal-core/src/meta.rs (hunk 2 above)
Comment (rowanG077) [reply irmaseal-core/src/meta.rs:26]: Possibly but I like this static size.

File: irmaseal-core/src/meta.rs (hunk 3 above)
Comment (rowanG077) [reply irmaseal-core/src/meta.rs:36]: Once we have const generics it should be possible but alas not yet.

File: irmaseal-core/src/meta.rs (hunk 1 above)
Comment (ivard) [reply irmaseal-core/src/meta.rs:31]: ArrayVec is an array wrapped in a package to make it possible to let it contain less than `size` values. It grows like a vector with the only difference that an array is used under water to store the values, so if you put more items in than the size of the underlying array you'll get an error.

In this case you don't need the growing functionality, since you exactly know how many items there will be. Therefore you can just say `iv: [u8; IVSIZE]`. The only question is then, can Serde parse the iv directly into a slice/array? Otherwise it is of course not possible to use a fixed array here.

**Comment by ivard** [comment]
The unit tests work fine, only when I try to compile it as part of `irmaseal-cli`, I get compile errors. Since `irmaseal-cli` is included in this repository, this compile error can be fixed within this PR. In `Cargo.tom` you can set `irmaseal-core` to `{ path = "../irmaseal-core" }`. Then you'll see the error.

**Comment by Wassasin** [comment]
Do not make `irmaseal-core` `std`. When you want to run this in an embedded context (for a secure element) or webassembly having this crate be `no-std` is essential.

Furthermore I think that having `json` in a file header is wasteful. There are currently no compact, non-alloc, no-std, serde-compatible serialization crates that have a nice format specification. If you drop the format specification requirement I recommend the [postcard](https://crates.io/crates/postcard) crate. I have exhaustively looked into this for other (embedded) projects. If it is a requirement, you could consider paying the postcard author to finish the specification document.

**Comment by rowanG077** [comment]
@Wassasin I moved to postcard and removed serde_json. Seems to work fine with the unit tests. Still todo is to move the CLI tool and pkg has not been migrated yet.

File: irmaseal-core/src/meta.rs (hunk 1 above)
Comment (rowanG077) [reply irmaseal-core/src/meta.rs:24]: `iv: [u8; IVSIZE]` doesn't work because rust array only work upto a specific length due to const generics not working yet,

File: irmaseal-core/src/stream/sealer.rs [hunk 4]
```diff
@@ -14,12 +16,13 @@ pub struct Sealer<'a, W: Writable> {
 
 impl<'a, W: Writable> Sealer<'a, W> {
     pub fn new<R: Rng + CryptoRng>(
-        i: &Identity,
+        i: Identity,
```
Comment (leonbotros) [inline irmaseal-core/src/stream/sealer.rs:19]: Why this change (and the other ones below as a consequence of this)?

File: irmaseal-pkg/Cargo.toml [hunk 5]
```diff
@@ -6,7 +6,7 @@ edition = "2018"
 license = "MIT"
 
 [dependencies]
-irmaseal-core = "0.1"
+irmaseal-core = { path = "../irmaseal-core" }
```
Comment (leonbotros) [inline irmaseal-pkg/Cargo.toml:9]: We should update these once we can publish

**Review by leonbotros — COMMENTED** [review]
Overall, its looks good. I left some questions though.

File: irmaseal-cli/src/decrypt.rs [hunk 6]
```diff
@@ -64,14 +68,15 @@ pub async fn exec(m: &ArgMatches<'_>) {
...
+        eprintln!("Disclosure successful, decrypting {}", input);
 
-        let mut o = o.unseal(&r.key.unwrap()).unwrap();
+        let unsealed =
+            crate::util::FileUnsealerRead::new(sealed.unseal(&metadata, &r.key.unwrap()).unwrap());
 
-        let mut of = crate::util::FileWriter::new(std::fs::File::create(output).unwrap());
-        o.write_to(&mut of).unwrap();
+        let mut tar = Archive::new(unsealed);
+        tar.unpack("./").unwrap();
```
[the remark this answers is not in the document]
Comment (rowanG077) [reply irmaseal-cli/src/decrypt.rs:77]: I don't think so.

File: irmaseal-cli/src/encrypt.rs [hunk 7]
```diff
@@ -38,32 +40,25 @@ pub async fn exec(m: &ArgMatches<'_>) {
...
-
-    let mut sealer = Sealer::new(&i, &parameters.public_key, &mut rng, &mut w).unwrap();
+    let input_path = Path::new(input);
+    let file_name_path = input_path.file_name().unwrap();
+    let file_name = file_name_path.to_str().unwrap();
 
-    use std::io::Read;
-    let mut src = std::fs::File::open(input).unwrap();
-    let mut buf = [0u8; 512];
+    let output = format!("{}.irma", file_name);
```
[the remark this answers is not in the document]
Comment (rowanG077) [reply irmaseal-cli/src/encrypt.rs:47]: Yes this would be fine with me.

**Comment by rowanG077** [comment]
@Wassasin Could you quickly look over the changes if everything is Oké with you?

File: irmaseal-cli/src/cli.yml [hunk 8]
```diff
@@ -35,10 +35,6 @@ subcommands:
         help: Sets the input file to use
         required: true
         index: 1
-    - OUTPUT:
```
Comment (ivard) [inline irmaseal-cli/src/cli.yml:38]: Is it really practical to remove this? If I have a file that is named exactly the same as a file within the tar, what happens then? Is it overwritten then or do I get an error?

**Review by leonbotros — APPROVED** [review]
